src/Elastic.Apm/Api/Context.cs

```diff
@@ -4,25 +4,29 @@
 
 using System;
 using System.Collections.Generic;
+using Elastic.Apm.Model;
 using Elastic.Apm.Report.Serialization;
 using Newtonsoft.Json;
 
 namespace Elastic.Apm.Api
 {
 	public class Context
```

Comment:
Should `Context` expose `SetLabel()` methods in order to set labels of different types, that `Transaction.SetLabel()` methods should call? In other words, are there usages where labels might be set on `Context` outside of the scope of a `Transaction`? There are some in unit tests

Reply:
> Should Context expose SetLabel() methods in order to set labels of different types, that Transaction.SetLabel() methods should call? In other words, are there usages where labels might be set on Context outside of the scope of a Transaction?

I don't think so - there are 2 use cases when you set labels for `Context`:

- with transactions and for those we have public `SetLabels()` methods
- for errors and in that case we also have a public API to pass labels: `ITransaction.CaptureError("Test Error", "TestMethod", new StackTrace().GetFrames(), labels: new Dictionary<string, Label> { { "stringLabel", "test" } });` . In case of errors the `Context` isn't exposed, because everything is filled when the error is captured on the fly - so no way to change an `Error.Context` after an error happened (expect with a filter).

I think exposing `Labels` on the `Context` wasn't a good idea because it just duplicates the API that we already have on `ITransaction` (and on the `CaptureError`/`CaptureException` methods) which is confusing: it's literally the same value with a duplicated API.

So overall, I think the cleanest would be to just keep the `.Labels` property with the obsolete attribute for backwards compatibility and then remove the whole Label thing from `Context`. 

> There are some in unit tests

I think I rewrote all of them to use the new public API (except the ones for backwards compatibility). In `SerializationTests` we have tests that new up a `Context` and serialize it and in those we use the new `internal InternalLabels` property. Those unit tests aren't real API use cases, in those cases there is no `Transaction` or `Error`, the tests only care about the serialization.

Reply:
got it, thanks